File: solutions/python/state-of-tic-tac-toe/1/state_of_tic_tac_toe.py
def gamestate(board):
    """
    Determines the state of a 3x3 tic-tac-toe game with specific error messages for Exercism.
    """
    # Flatten the board to count Xs and Os
    flat_board = "".join(board)
    count_x = flat_board.count('X')
    count_o = flat_board.count('O')
    
    # Validate turn order counts with specific Exercism error messages
    if count_x > count_o + 1:
        raise ValueError("Wrong turn order: X went twice")
    if count_o > count_x:
        raise ValueError("Wrong turn order: O started")
    
    def check_win(player):
        # Rows
        for row in board:
            if all(cell == player for cell in row):
                return True
        # Columns
        for col in range(3):
            if all(board[row][col] == player for row in range(3)):
                return True
        # Diagonals
        if all(board[i][i] == player for i in range(3)):
            return True
        if all(board[i][2-i] == player for i in range(3)):
            return True
        return False
    
    x_wins = check_win('X')
    o_wins = check_win('O')
    
    # Check for impossible state: both players winning or continued play
    if x_wins and o_wins:
        raise ValueError("Impossible board: game should have ended after the game was won")
    
    if x_wins:
        # If X wins, they must have exactly one more mark than O
        if count_x != count_o + 1:
            raise ValueError("Impossible board: game should have ended after the game was won")
        return "win"
        
    if o_wins:
        # If O wins, mark counts must be equal
        if count_x != count_o:
            raise ValueError("Impossible board: game should have ended after the game was won")
        return "win"
    
    # Final state check
    if ' ' not in flat_board:
        return "draw"
    
    return "ongoing"

File: solutions/typescript/state-of-tic-tac-toe/1/state-of-tic-tac-toe.ts
export type State = 'ongoing' | 'draw' | 'win';

export const gamestate = (board: string[]): State => {
  const grid = board.map(row => row.split(''));
  let xCount = 0;
  let oCount = 0;

  // 1. Count marks
  for (const row of grid) {
    for (const cell of row) {
      if (cell === 'X') xCount++;
      if (cell === 'O') oCount++;
    }
  }

  // 2. Validate turn order (X starts)
  if (oCount > xCount) {
    throw new Error('Wrong turn order: O started');
  }
  if (xCount > oCount + 1) {
    throw new Error('Wrong turn order: X went twice');
  }

  // 3. Function to check winners
  const checkWin = (p: string): boolean => {
    const lines = [
      // Rows
      [grid[0][0], grid[0][1], grid[0][2]],
      [grid[1][0], grid[1][1], grid[1][2]],
      [grid[2][0], grid[2][1], grid[2][2]],
      // Columns
      [grid[0][0], grid[1][0], grid[2][0]],
      [grid[0][1], grid[1][1], grid[2][1]],
      [grid[0][2], grid[1][2], grid[2][2]],
      // Diagonals
      [grid[0][0], grid[1][1], grid[2][2]],
      [grid[0][2], grid[1][1], grid[2][0]],
    ];
    return lines.some(line => line.every(cell => cell === p));
  };

  const xWins = checkWin('X');
  const oWins = checkWin('O');

  // 4. Validate if game continued after win
  if (xWins && oWins) {
    throw new Error('Impossible board: game should have ended after the game was won');
  }
  
  // If X won, X must have one more move than O (X was last to move)
  if (xWins && xCount === oCount) {
    throw new Error('Impossible board: game should have ended after the game was won');
  }

  // If O won, X and O must have equal moves (O was last to move)
  if (oWins && xCount > oCount) {
    throw new Error('Impossible board: game should have ended after the game was won');
  }

  // 5. Return final state
  if (xWins || oWins) return 'win';
  if (xCount + oCount === 9) return 'draw';
  
  return 'ongoing';
};